### PR TITLE
[watermarking] context deduplication during generation

### PR DESCRIPTION
**Pull request by TQCB**
## Purpose
Text watermarking in vLLM (see https://github.com/vllm-project/vllm/pull/54053) currently uses a short context (4 tokens, by default) before each generated token to create the pseudorandom noise that we use to watermark its sampling. It is possible for multiple positions in a sequence to have the same context, and thus have the same watermarking bias applied. This makes gumbel watermarking single-token non-distortionary instead of single-sequence non-distortionary: expectation of token distribution, over all keys, is unchanged, but expectation of sequence distributions, over all keys, is changed. In practice, this leads to:
- a reduced diversity of generations within a sequence, and
- increased likelihood of poor behaviour such a repetition and infinite generations

A solution to this is context masking, or context deduplication: if a context has already been watermarked in a sequence, we do not watermark it again. This allows for single-sequence non-distortion (see section G.3 of [SynthID-Text supplementary materials](https://media.springernature.com/original/springer-static/esm/art%3A10.1038%2Fs41586-024-08025-4/MediaObjects/41586_2024_8025_MOESM1_ESM.pdf)). An initial instinct may be that reducing the amount of tokens we watermark within a sequence will reduce the strength of our watermarking signal, but we actually already deduplicate contexts during detection to avoid biasing repetitions: we only score the first instance of a repeated context, matching what we propose for generation here. In practice however, we may imagine a case where a user will manipulate watermarked text to excluse an initial watermarking of a repeated context, only keeping the subsequent unwatermarked repetitions: this would indeed lead to a reduced watermarking signal, but this is already expected with a tampered text, and we propose that this is acceptable if it allows for increased guarantees of non-distortion.

## Test Plan
Measuring impact, using standard models and methods, on:
- degenerate generations (repetition, infinite generations)
- output length, intra-sequence diversity
- output quality
- inference performance

## Test Result
Results for commit `42908fd643fb46744f78937dc39034b02c99b172` on 1x H100

#### Microbenchmark
- ms per sample step
- mean of 100 + 20 warmup steps

| Batch | History tokens | Unmasked (ms/step) | Masked (ms/step) | Delta ms | Delta % |
|---|---|---|---|---|---|
| 1 | 2,048 | 0.036 | 0.110 | +0.073 | +201.2% |
| 1 | 8,192 | 0.036 | 0.111 | +0.075 | +208.7% |
| 1 | 32,768 | 0.037 | 0.110 | +0.073 | +197.5% |
| 32 | 2,048 | 0.041 | 0.110 | +0.069 | +166.6% |
| 32 | 8,192 | 0.041 | 0.110 | +0.069 | +165.6% |
| 32 | 32,768 | 0.042 | 0.111 | +0.069 | +166.8% |
| 256 | 2,048 | 0.260 | 0.351 | +0.092 | +35.3% |
| 256 | 8,192 | 0.260 | 0.482 | +0.222 | +85.5% |
| 256 | 32,768 | 0.260 | 0.483 | +0.223 | +86.0% |
| 1,024 | 2,048 | 1.003 | 1.280 | +0.276 | +27.6% |
| 1,024 | 8,192 | 1.004 | 1.793 | +0.790 | +78.7% |
| 1,024 | 32,768 | 1.003 | 1.792 | +0.789 | +78.6% |

#### E2E
Qwen3.5-2B

| Model | Batch | Output tokens | Unmasked tok/s | Masked tok/s | Δ throughput |
|---|---|---|---|---|---|
| Qwen3.5-2B | 1 | 2,048 | 401.4 | 408.9 | +1.88% |
| Qwen3.5-2B | 32 | 2,048 | 9,867.9 | 9,318.3 | -5.57% |
| Qwen3.5-2B | 1 | 8,192 | 402.9 | 422.7 | +4.92% |
| Qwen3.5-2B | 32 | 8,192 | 8,487.2 | 8,675.3 | +2.22% |

Qwen3.5-27B (mean of 3)

| Model | Batch | Output tokens | Unmasked tok/s | Masked tok/s | Δ throughput |
|---|---|---|---|---|---|
| Qwen3.5-27B | 1 | 512 | 48.2 | 48.2 | +0.02% |
| Qwen3.5-27B | 32 | 512 | 1,248.9 | 1,246.7 | -0.18% |
| Qwen3.5-27B | 1 | 2,048 | 48.3 | 48.3 | -0.05% |
| Qwen3.5-27B | 32 | 2,048 | 1,233.9 | 1,231.5 | -0.19% |
| Qwen3.5-27B | 1 | 8,192 | 48.2 | 48.2 | -0.05% |

### Quality & Diversity
Qwen3.5-2B T=1.0

| Metric | No watermark | Watermark, unmasked | Watermark, masked |
|---|---|---|---|
| Generations | 300 | 300 | 300 |
| Generations reaching the 1,024-token cap | 19 (6.3%) | 107 (35.7%) | 12 (4.0%) |
| Repeated 4-token context rate (micro) | 25.37% | 74.94% | 17.99% |
| Mean completion length (tokens) | 211.4 | 444.8 | 188.9 |
| Answer accuracy | 29.7% | 25.7% | 29.0% |

image

image

### Detectability
Detectability on the GSM8K responses

| Metric | No watermark | Watermark, unmasked | Watermark, masked |
|---|---|---|---|
| Positive at p ≤ 0.01 | 8/300 (2.7%) | 262/300 (87.3%) | 260/300 (86.7%) |
| Mean p-value | 0.51183 | 0.02107 | 0.01689 |
| Mean scored tokens | 157.8 | 111.5 | 154.9 |

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/watermark.py [hunk 1]
```diff
@@ -42,6 +42,88 @@
...
+            historical_pos = previous_output_pos + offset - CONTEXT_WIDTH
+            historical_token = tl.load(
+                all_token_ids_ptr
+                + safe_req_idx * all_token_ids_stride
+                + prompt_len
+                + tl.maximum(historical_pos, 0),
+                mask=valid_req & (historical_pos >= 0),
+                other=-1,
+            )
+            context_token = tl.load(contexts_ptr + row * context_stride + offset)
```
Comment (tomasruizt) [inline vllm/v1/worker/gpu/sample/watermark.py]: the context token can be loaded outside the loop right? The compiler might hoist it out, but its better if we do. Edit: Just saw that it depends on the `offset`, too.

File: vllm/v1/worker/gpu/sample/watermark.py [hunk 2]
```diff
@@ -42,6 +42,88 @@
...
+            current_context = tuple(contexts[row].tolist())
+            for token_id in output_tokens:
+                if tuple(prefix[-contexts.shape[-1] :]) == current_context:
+                    repeated[row] = True
+                    break
+                prefix.append(token_id)
+        return repeated
+
+    repeated = torch.empty(len(req_indices), dtype=torch.bool, device=contexts.device)
+    _repeated_context_mask_kernel[(len(req_indices),)](
```
Comment (tomasruizt) [inline vllm/v1/worker/gpu/sample/watermark.py:184]: Here we only launch a few programs (e.g. 10 if there are 10 requests), and have a long loop per request (e.g. 100k in long context). This underutilizes the GPU. Instead, we could split up the work on many programs per request, and if any program finds a duplicate, they can set the flag for their request to True using an atomic operation.

File: tests/watermarking/test_watermarking.py [hunk 3]
```diff
@@ -123,6 +129,169 @@ def sample(self, logits, contexts, random_sample):
...
+        total_lens,
+        contexts,
+    )
+
+    assert torch.equal(repeated, torch.tensor([True, False, False]))
+
+
+def _repeated_context_inputs(
+    context_width: int, device: str = "cpu"
+) -> tuple[torch.Tensor, ...]:
```
Comment (tomasruizt) [inline tests/watermarking/test_watermarking.py:495]: A comment would help clarify what sort of inputs this function constructs.

File: vllm/v1/worker/gpu/sample/watermark.py [hunk 4]
```diff
@@ -42,6 +42,88 @@
...
+            total_len = int(total_lens[req_idx])
+            output_tokens = all_token_ids[req_idx, prompt_len:total_len].tolist()
+            prefix = [-1] * contexts.shape[-1]
+            current_context = tuple(contexts[row].tolist())
+            for token_id in output_tokens:
+                if tuple(prefix[-contexts.shape[-1] :]) == current_context:
+                    repeated[row] = True
+                    break
+                prefix.append(token_id)
+        return repeated
```
Comment (tomasruizt) [inline vllm/v1/worker/gpu/sample/watermark.py]: perhaps this can be put in a fn named `repeated_context_mask_cpu()`

**Comment by simon-veitner-redhat** [comment]
## `_repeated_context_mask_kernel`: serial history rescan, and a block-parallel replacement

### Problem

The kernel runs one Triton program per sampled row with a scalar `while` loop over every previous output position, so each decode step does O(output_len · W) serial work per row with no parallelism inside the row. `all_token_ids` is a UVA view of pinned host memory, so every scalar load also crosses PCIe. Latency is set by the longest row and is nearly flat in batch size, so one long request taxes every co-scheduled request.

Mask kernel alone, ms per decode step, H200, W=4, production memory layout:

| output_len | rows=1 | rows=256 | rows=1024 |
|---:|---:|---:|---:|
| 2048 | 0.47 | 0.52 | 0.74 |
| 8192 | 1.79 | 2.00 | 2.88 |
| 32768 | 7.10 | 7.95 | 11.41 |

For scale, `philox_gumbel_sample` over the full 151936 vocab costs 0.04 ms at 1 row and 0.62 ms at 1024 rows.

### Fix: block-parallel scan with identical semantics

Each iteration tests `BLOCK` candidate positions at once and reduces with `tl.max`. Output is bit-identical to the current kernel for W in {1, 2, 4, 16}, including padding rows (`req_idx < 0`) and `output_len == 0`. Launch with `BLOCK=512`; the wrapper is otherwise unchanged.

```python
@triton.jit
def _repeated_context_mask_kernel(
    output_ptr,
    all_token_ids_ptr,
    all_token_ids_stride,
    req_indices_ptr,
    prompt_lens_ptr,
    total_lens_ptr,
    contexts_ptr,
    context_stride,
    CONTEXT_WIDTH: tl.constexpr,
    BLOCK: tl.constexpr,
):
    row = tl.program_id(0).to(tl.int64)
    req_idx = tl.load(req_indices_ptr + row).to(tl.int64)
    valid = req_idx >= 0
    safe = tl.maximum(req_idx, 0)
    plen = tl.load(prompt_lens_ptr + safe, mask=valid, other=0).to(tl.int64)
    tlen = tl.load(total_lens_ptr + safe, mask=valid, other=0).to(tl.int64)
    olen = tlen - plen
    base_ptr = all_token_ids_ptr + safe * all_token_ids_stride + plen

    repeated = False
    for start in range(0, olen, BLOCK):
        # Each lane tests one previous output position p in [start, start + BLOCK).
        pos = start + tl.arange(0, BLOCK)
        in_range = pos < olen
        match = in_range & valid
        for offset in tl.static_range(CONTEXT_WIDTH):
            # Token at output position p - CONTEXT_WIDTH + offset, -1 before output start.
            hp = pos + offset - CONTEXT_WIDTH
            tok = tl.load(
                base_ptr + tl.maximum(hp, 0),
                mask=in_range & valid & (hp >= 0),
                other=-1,
            )
            ctx = tl.load(contexts_ptr + row * context_stride + offset)
            match &= tok == ctx
        repeated |= tl.max(match.to(tl.int32)) > 0
    tl.store(output_ptr + row, repeated)
```

Two small things worth folding in: hoist the `ctx` loads out of the `start` loop (loop invariant, and the compiler does not hoist them), and add the `contexts.stride(-1) != 1` guard that `philox_gumbel_sample` already has, since both kernels assume a unit last-dimension stride.

### Kernel microbenchmark

ms per step, H200, production memory layout:

| output_len | rows | W | current | blocked | speedup |
|---:|---:|---:|---:|---:|---:|
| 8192 | 1 | 4 | 1.71 | 0.031 | 56x |
| 8192 | 256 | 4 | 1.92 | 0.18 | 11x |
| 8192 | 1024 | 4 | 2.87 | 0.72 | 4x |
| 8192 | 1024 | 16 | 10.13 | 0.82 | 12x |
| 32768 | 1 | 4 | 6.80 | 0.11 | 64x |
| 32768 | 256 | 4 | 7.62 | 0.70 | 11x |
| 32768 | 1024 | 4 | 11.39 | 2.83 | 4x |
| 32768 | 1024 | 16 | 40.50 | 3.21 | 13x |

What remains at large batch is PCIe bandwidth from scanning UVA memory; the same blocked kernel on device memory is another 10 to 40x lower. An incremental per-request context-hash array, appended to in `_post_update_kernel`, would make the per-step check O(1) and remove the PCIe reads entirely, but that is a larger change and the blocked scan is enough for this PR.

### End to end

Qwen3.5-2B, 8192-token outputs, `ignore_eos`, temperature 1.0, CUDA graphs on, mean of 2 runs:

| requests | dedup off | current kernel | blocked kernel |
|---:|---:|---:|---:|
| 1 | 433 tok/s, 2.31 ms/step | 330 tok/s, 3.03 ms/step (-24%) | 421 tok/s, 2.38 ms/step (-3%) |
| 32 | 9746 tok/s, 3.28 ms/step | 7706 tok/s, 4.15 ms/step (-21%) | 9054 tok/s, 3.54 ms/step (-7%) |

With the current kernel the regression grows with output length: -7% tok/s at 1500-token outputs, -18% at 6000, -21% at 8192. The residual 3 to 7% with the blocked kernel is the always-on fallback `gumbel_sample` (below) plus the UVA reads.

### Accuracy

GSM8K test set, 300 problems, Qwen3.5-2B, temperature 1.0, max 768 tokens, key 42. The blocked kernel produces identical output to the current one, so this covers both.

| mode | accuracy | hit token cap | cap hits that are repetition loops | detected at p 0.01 |
|---|---:|---:|---:|---:|
| unwatermarked | 67.0% | 27.0% | 3 | 1.3% |
| watermark, dedup off | 68.7% | 22.0% | 10 | 71.0% |
| watermark, dedup on | 67.3% | 22.3% | 2 | 72.7% |

Accuracy is flat within noise. Dedup removes the repetition loops the single-key sampler introduces (10 of 66 cap hits down to 2, back at the unwatermarked baseline of 3) and detection improves slightly.

### Related, independent of the kernel

With `deduplicate_contexts=True` the `repeated_contexts is not None` branch in `_sample_random` is always taken, so the fallback `gumbel_sample` plus `torch.where` launch on every step even when no row is masked (about +0.1 ms at 1 row, +0.95 ms at 1024 rows). Folding masked rows into `philox_gumbel_sample` (seed/pos noise for those rows) would remove the second full-vocab kernel.

---

*Disclaimer: AI was used to create this comment*

File: vllm/v1/worker/gpu/sample/watermark.py (hunk 2 above)
Comment (tomasruizt) [reply vllm/v1/worker/gpu/sample/watermark.py:184]: @simon-veitner-redhat

**Comment by simon-veitner-redhat** [comment]
within `vllm/config/watermarking.py:41-46 ` adjust 

```python
        if self.algorithm == "gumbel":
            logger.warning_once(
                "Single-key Gumbel-max watermarking may increase the frequency of "
                "degenerate generations, including repetition loops.",
                scope="global",
            )
```

to something like

```python
        if self.algorithm == "gumbel" and not self.deduplicate_contexts:
            logger.warning_once(
                "Single-key Gumbel-max watermarking with deduplicate_contexts=False "
                "may increase the frequency of degenerate generations, including "
                "repetition loops; keep deduplicate_contexts=True to mitigate this.",
                scope="global",
            )
```

adjust corresponding test

**Comment by LucasWilkinson** [comment]
Closing as superseded by #56122, which was branched on top of this PR and carried all of its commits (context deduplication config, repeated-context mask kernel, sampler wiring, detector changes, docs, and tests) into main when it was squash-merged.

After merging latest main into this branch, the only remaining difference is the `WatermarkContextScope` → `WatermarkDeduplicationScope` type alias rename from the last commit.

@TQCB if I have this wrong and there is something here that did not land via #56122, please reopen.
